In [ ]:
import pandas as pd
import numpy as np

In [ ]:
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [ ]:
file_path = "/content/track2_identity_asset_master.csv"

In [ ]:
df = pd.read_csv(file_path)
print("Shape:", df.shape)
df.head(10)

Shape: (3090, 12)


,user_id,username,full_name,department,role,location,hostname,device_id,status,hire_date,termination_date,manager_username
0,EMP12741,OMKAAR.CHANA52,Omkaar Chana,Supply Chain,Administrator,Branch Office,LPT-12741,NaN,Working,1736578363,not available,manager780
1,EMP11307,ethan.narasimhan97,Ethan Narasimhan,LEGAL,Manager,Head Office,VDR-11307.corp.local,dev42831,Live,18/05/2026 23:12,NaN,manager881
2,EMP11601,ethan.choudhry47,Ethan Choudhry,Marketing,Director,NaN,WS-11601,dev40763,Live,14/01/2023 09:22,NaN,manager809
3,12621,vedant_panchal60,Vedant Panchal,Support,Analyst,Head_Office,LPT_12621,DEV-95371,Enabled,1656531399,NaN,manager886
4,EMP12029,gayathri.mohanty65,Gayathri Mohanty,RnD,Executive,work from home,LPT-12029,dev 51958,A,2021/09/01,NaN,manager867
5,EMP11950,GARIMA.CHAUDHARY44,Garima Chaudhary,Call Center,Executive,NaN,LPT-11950,dev13049,Live,2023-12-30T17:13:00,NaN,manager521
6,EMP 12718,vrishti.dewan18,Vrishti Dewan,Compliance,Officer,warehouse,lpt-12718,DEV-46814,Working,2021-06-21 08:34:45,NaN,manager544
7,EMP10184,NaN,Varsha Mall,IT,Director,Remote,WS-10184,dev 12128,A,28/07/2020 21:36,NaN,MANAGER850
8,EMP12400,peter.sama77,Peter Sama,information tech,Specialist,work from home,VDR-12400,NaN,Enabled,06-13-2026 07:10:36 PM,NaN,manager829
9,EMP 10194,ayush.dua87,Ayush Dua,RnD,Analyst,BRANCH OFFICE,LPT-10194,DEV-40932,ACTIVE,2023-02-12 03:25:50,not available,manager680


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3090 entries, 0 to 3089
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   user_id           3090 non-null   object
 1   username          2965 non-null   object
 2   full_name         3090 non-null   object
 3   department        3090 non-null   object
 4   role              3090 non-null   object
 5   location          2516 non-null   object
 6   hostname          2916 non-null   object
 7   device_id         2507 non-null   object
 8   status            3090 non-null   object
 9   hire_date         2753 non-null   object
 10  termination_date  1265 non-null   object
 11  manager_username  2436 non-null   object
dtypes: object(12)
memory usage: 289.8+ KB


In [ ]:
missing = pd.DataFrame({
    "missing_count": df.isna().sum(),
    "missing_%": (df.isna().mean() * 100).round(2)
})

missing.sort_values("missing_%", ascending=False)

,missing_count,missing_%
termination_date,1825,59.06
manager_username,654,21.17
device_id,583,18.87
location,574,18.58
hire_date,337,10.91
hostname,174,5.63
username,125,4.05
user_id,0,0.00
department,0,0.00
full_name,0,0.00


In [ ]:
print("Duplicate rows:", df.duplicated().sum())

print("Duplicate user_ids:", df["user_id"].duplicated().sum())

print("Unique user_ids:", df["user_id"].nunique())

Duplicate rows: 90
Duplicate user_ids: 90
Unique user_ids: 3000


In [ ]:
for col in ["department", "location", "status"]:
    print("\n================", col, "================")
    print(df[col].value_counts(dropna=False))


================ department ================
department
Ops Team                    84
People Team                 81
information tech            71
sales dept                  70
hr dept                     70
finance dept                69
Ops                         69
RnD                         69
Human Resource              69
Operations                  68
Purchase                    68
MKT                         68
Legal Dept                  68
Support                     67
IT                          66
FINANCE                     66
Business Sales              65
R&D                         65
PURCH                       65
IT Support                  63
procurement team            63
IT Dept                     63
Accounts                    62
Finance                     62
HR                          62
Customer Support            62
Supply Chain                62
legal                       61
Innovation                  59
Information Technology      57
RD           

In [ ]:
for col in ["user_id", "hostname", "device_id", "manager_username"]:
    print("\n================", col, "================")
    print("Unique:", df[col].nunique(dropna=False))
    print(df[col].dropna().astype(str).head(20).tolist())


================ user_id ================
Unique: 3000
['EMP12741', 'EMP11307', 'EMP11601', '12621', 'EMP12029', 'EMP11950', 'EMP 12718', 'EMP10184', 'EMP12400', 'EMP 10194', 'emp12798', '11195', 'emp11406', 'EMP10314', 'EMP10102', 'EMP 11437', 'emp_12959', 'EMP11221', 'EMP 12935', 'emp_10705']

================ hostname ================
Unique: 2830
['LPT-12741', 'VDR-11307.corp.local', 'WS-11601', 'LPT_12621', 'LPT-12029', 'LPT-11950', 'lpt-12718', 'WS-10184', 'VDR-12400', 'LPT-10194', 'WS-12798', 'VDR-11195', 'WS-11406', 'VDR-10314', 'WS-10102', 'WS-11437', 'VDR-12959', 'WS-11221', 'WS-12935', 'LPT-10705.corp.local']

================ device_id ================
Unique: 2422
['dev42831', 'dev40763', 'DEV-95371', 'dev 51958', 'dev13049', 'DEV-46814', 'dev 12128', 'DEV-40932', 'DEV31013', 'DEV99205', 'DEV78076', 'DEV-48986', 'DEV-54548', 'DEV76029', 'DEV19450', 'DEV32769', 'dev47829', 'DEV-92483', 'DEV-46572', 'dev 94430']

================ manager_username ================
Unique: 11

In [ ]:
df['hire_date'].map(type).value_counts()


,count
hire_date,
<class 'str'>,2753
<class 'float'>,337


In [ ]:
print(df["hire_date"].value_counts(dropna=False).head(30))
print("\nTermination dates:")
print(df["termination_date"].value_counts(dropna=False).head(30))


hire_date
NaN                       337
19/04/2023                  4
01/09/2024                  3
23/10/2023 22:52            2
2024-07-03 12:34:35         2
2022-06-25 06:02:35         2
21-Apr-2025 01:20:59        2
06-24-2020 08:36:29 AM      2
2020/11/16                  2
2021/08/29                  2
1688769300                  2
2020/03/01                  2
21/05/2021                  2
2025-10-19 23:21:55         2
05-May-2023 19:22:01        2
2022/06/30                  2
2025-02-12 17:19:32         2
01/07/2023                  2
04/01/2026 18:47            2
2020-10-04 23:45:27         2
2021/10/16                  2
12-05-2025 06:46:12 AM      2
07-21-2025 04:04:10 AM      2
1646993435                  2
1643561219                  2
2024-07-07T08:13:23         2
27/09/2021 22:58            2
2020/11/10                  2
2024-07-23 23:08:34         2
2020-01-27T12:19:55         2
Name: count, dtype: int64

Termination dates:
termination_date
NaN                       1

In [ ]:
# Standardizing user_id

df["user_id_clean"] = (
    df["user_id"]
    .astype(str)
    .str.upper()
    .str.replace(r"[^A-Z0-9]", "", regex=True)
)

df[["user_id", "user_id_clean"]].head(20)

,user_id,user_id_clean
0,EMP12741,EMP12741
1,EMP11307,EMP11307
2,EMP11601,EMP11601
3,12621,12621
4,EMP12029,EMP12029
5,EMP11950,EMP11950
6,EMP 12718,EMP12718
7,EMP10184,EMP10184
8,EMP12400,EMP12400
9,EMP 10194,EMP10194


In [ ]:
print("Original unique IDs:", df["user_id"].nunique())
print("Clean unique IDs:", df["user_id_clean"].nunique())

print("\nDuplicate cleaned IDs:")
print(
    df[df["user_id_clean"].duplicated(keep=False)]
    [["user_id", "user_id_clean", "username", "hostname", "device_id"]]
    .sort_values("user_id_clean")
    .head(50)
)

Original unique IDs: 3000
Clean unique IDs: 3000

Duplicate cleaned IDs:
        user_id user_id_clean                username              hostname  \
1674      10063         10063               IRA.TARA4                   NaN   
79        10063         10063               IRA.TARA4                   NaN   
1108      11190         11190         yuvraj.mistry78             LPT-11190   
2794      11190         11190         yuvraj.mistry78             LPT-11190   
11        11195         11195      christopher.toor37             VDR-11195   
1455      11195         11195      christopher.toor37             VDR-11195   
714       11359         11359         gaurika.patla96             LPT-11359   
1755      11359         11359         gaurika.patla96             LPT-11359   
2736      11491         11491          zarna.mandal21             LPT-11491   
525       11491         11491          zarna.mandal21             LPT-11491   
1216   EMP10014      EMP10014            ridhi.saini2     

In [ ]:
# Check whether duplicate user_id records are completely identical

duplicate_users = df[df["user_id_clean"].duplicated(keep=False)].copy()

duplicate_users = duplicate_users.sort_values("user_id_clean")

duplicate_users.head(20)

,user_id,username,full_name,department,role,location,hostname,device_id,status,hire_date,termination_date,manager_username,user_id_clean
1674,10063,IRA.TARA4,Ira Tara,People Team,Director,branch office,NaN,DEV-33819,ACTIVE,NaN,NaN,manager565,10063
79,10063,IRA.TARA4,Ira Tara,People Team,Director,branch office,NaN,DEV-33819,ACTIVE,NaN,NaN,manager565,10063
1108,11190,yuvraj.mistry78,Yuvraj Mistry,Support,Consultant,DATA CENTER,LPT-11190,NaN,Enabled,21-Apr-2025 01:20:59,NaN,NaN,11190
2794,11190,yuvraj.mistry78,Yuvraj Mistry,Support,Consultant,DATA CENTER,LPT-11190,NaN,Enabled,21-Apr-2025 01:20:59,NaN,NaN,11190
11,11195,christopher.toor37,Christopher Toor,legal,Director,work from home,VDR-11195,DEV99205,Enabled,02-13-2023 09:32:23 AM,NaN,manager739,11195
1455,11195,christopher.toor37,Christopher Toor,legal,Director,work from home,VDR-11195,DEV99205,Enabled,02-13-2023 09:32:23 AM,NaN,manager739,11195
714,11359,gaurika.patla96,Gaurika Patla,Human Resources,Executive,Work From Home,LPT-11359,dev 66547,Enabled,02/06/2021 15:37,NaN,NaN,11359
1755,11359,gaurika.patla96,Gaurika Patla,Human Resources,Executive,Work From Home,LPT-11359,dev 66547,Enabled,02/06/2021 15:37,NaN,NaN,11359
2736,11491,zarna.mandal21,Zarna Mandal,Sales,Trainee,NaN,LPT-11491,dev 10902,Active,2025-10-19 23:21:55,NaN,manager987,11491
525,11491,zarna.mandal21,Zarna Mandal,Sales,Trainee,NaN,LPT-11491,dev 10902,Active,2025-10-19 23:21:55,NaN,manager987,11491


In [ ]:
# For each duplicated user_id, check how many distinct values
# exist in each column

duplicate_users.groupby("user_id_clean").nunique()

,user_id,username,full_name,department,role,location,hostname,device_id,status,hire_date,termination_date,manager_username
user_id_clean,,,,,,,,,,,,
10063,1,1,1,1,1,1,0,1,1,0,0,1
11190,1,1,1,1,1,1,1,0,1,1,0,0
11195,1,1,1,1,1,1,1,1,1,1,0,1
11359,1,1,1,1,1,1,1,1,1,1,0,0
11491,1,1,1,1,1,0,1,1,1,1,0,1
EMP10014,1,1,1,1,1,1,1,1,1,1,1,1
EMP10044,1,1,1,1,1,0,1,1,1,1,0,1
EMP10051,1,1,1,1,1,0,1,1,1,1,1,1
EMP10052,1,1,1,1,1,0,1,1,1,1,1,1


In [ ]:
# How many rows are exact duplicates?

print("Exact duplicate rows:", df.duplicated().sum())

Exact duplicate rows: 90


In [ ]:
print("Duplicate rows:", df.duplicated().sum())
print("Duplicate user IDs:", df["user_id_clean"].duplicated().sum())


Duplicate rows: 90
Duplicate user IDs: 90


In [ ]:
# ============================================
# REMOVE EXACT DUPLICATES
# ============================================

before = len(df)

df = df.drop_duplicates().copy()

after = len(df)

print("Rows before:", before)
print("Rows after:", after)
print("Duplicate rows removed:", before - after)

Rows before: 3090
Rows after: 3000
Duplicate rows removed: 90


In [ ]:
print("Remaining duplicate rows:", df.duplicated().sum())
print("Unique users:", df["user_id_clean"].nunique())

Remaining duplicate rows: 0
Unique users: 3000


In [ ]:
df["department"].value_counts().sort_index()

,count
department,
Accounts,61
Brand Team,54
Business Sales,65
CS,56
Call Center,52
Compliance,51
Customer Support,62
FINANCE,66
Fin,49


In [ ]:
department_map = {
    # Finance
    "Accounts": "Finance",
    "FINANCE": "Finance",
    "Fin": "Finance",
    "Finance": "Finance",
    "finance dept": "Finance",

    # Human Resources
    "HR": "Human Resources",
    "Human Resource": "Human Resources",
    "Human Resources": "Human Resources",
    "hr dept": "Human Resources",

    # Information Technology
    "IT": "IT",
    "IT Dept": "IT",
    "IT Support": "IT",
    "Information Technology": "IT",
    "information tech": "IT",

    # Legal
    "LEGAL": "Legal",
    "Legal": "Legal",
    "Legal Dept": "Legal",
    "legal": "Legal",

    # Marketing
    "MKT": "Marketing",
    "Marketing": "Marketing",
    "Mktg": "Marketing",
    "marketing dept": "Marketing",

    # Operations
    "OPERATIONS": "Operations",
    "Operations": "Operations",
    "Ops": "Operations",
    "Ops Team": "Operations",
    "operations dept": "Operations",

    # Procurement
    "PURCH": "Procurement",
    "Procurement": "Procurement",
    "Purchase": "Procurement",
    "procurement team": "Procurement",

    # Research & Development
    "R&D": "R&D",
    "RD": "R&D",
    "Research and Development": "R&D",
    "RnD": "R&D",

    # Sales
    "SALES": "Sales",
    "Sales": "Sales",
    "Business Sales": "Sales",
    "sales dept": "Sales",
    "sales team": "Sales",

    # Customer Support
    "CS": "Customer Support",
    "Customer Support": "Customer Support",
    "Support": "Customer Support",
    "customer care": "Customer Support",
    "Call Center": "Customer Support",

    # Supply Chain
    "Supply Chain": "Supply Chain",

    # People
    "People Team": "People",

    # Compliance
    "Compliance": "Compliance",

    # Brand
    "Brand Team": "Brand",

    # Innovation
    "Innovation": "Innovation"
}

df["department_clean"] = df["department"].map(department_map)

In [ ]:
print("Unmapped departments:")
print(df.loc[df["department_clean"].isna(), "department"].unique())

Unmapped departments:
[]


In [ ]:
df["department_clean"].value_counts().sort_index()

,count
department_clean,
Brand,54
Compliance,51
Customer Support,288
Finance,302
Human Resources,249
IT,307
Innovation,58
Legal,231
Marketing,223


In [ ]:
##location

df["location"].value_counts(dropna=False).sort_index()


,count
location,
BRANCH OFFICE,90
Branch Office,95
Branch_Office,92
DATA CENTER,106
Data Center,80
Data_Center,89
HEAD OFFICE,73
Head Office,90
Head_Office,89


In [ ]:
location_map = {
    # Branch Office
    "BRANCH OFFICE": "Branch Office",
    "Branch Office": "Branch Office",
    "Branch_Office": "Branch Office",
    "branch office": "Branch Office",

    # Data Center
    "DATA CENTER": "Data Center",
    "Data Center": "Data Center",
    "Data_Center": "Data Center",
    "data center": "Data Center",

    # Head Office
    "HEAD OFFICE": "Head Office",
    "Head Office": "Head Office",
    "Head_Office": "Head Office",
    "head office": "Head Office",

    # Regional Office
    "REGIONAL OFFICE": "Regional Office",
    "Regional Office": "Regional Office",
    "Regional_Office": "Regional Office",
    "regional office": "Regional Office",

    # Remote
    "REMOTE": "Remote",
    "Remote": "Remote",
    "remote": "Remote",

    # Warehouse
    "WAREHOUSE": "Warehouse",
    "Warehouse": "Warehouse",
    "warehouse": "Warehouse",

    # Work From Home
    "WORK FROM HOME": "Work From Home",
    "Work From Home": "Work From Home",
    "Work_From_Home": "Work From Home",
    "work from home": "Work From Home"
}

df["location_clean"] = df["location"].map(location_map)

In [ ]:
print("Unmapped locations:")
print(df.loc[
    df["location"].notna() & df["location_clean"].isna(),
    "location"
].unique())

Unmapped locations:
[]


In [ ]:
df["location_clean"].value_counts(dropna=False).sort_index()

,count
location_clean,
Branch Office,367
Data Center,358
Head Office,337
Regional Office,320
Remote,351
Warehouse,326
Work From Home,377
NaN,564


In [ ]:
#status


In [ ]:
df["status"].value_counts(dropna=False).sort_index()

,count
status,
A,402
ACTIVE,412
Active,361
Blocked,37
D,45
DISABLED,59
Deactivated,57
Disabled,41
Enabled,378


In [ ]:
status_map = {
    # Active
    "A": "Active",
    "ACTIVE": "Active",
    "Active": "Active",
    "Enabled": "Active",
    "Live": "Active",
    "Working": "Active",

    # Disabled
    "Blocked": "Disabled",
    "D": "Disabled",
    "DISABLED": "Disabled",
    "Deactivated": "Disabled",
    "Disabled": "Disabled",

    # Inactive / exited
    "Exited": "Inactive",
    "L": "Inactive",
    "LEFT": "Inactive",
    "Left": "Inactive",
    "LWP": "Inactive",
    "Leave": "Inactive",
    "Resigned": "Inactive",
    "Terminated": "Inactive",

    # Temporarily away
    "ON_LEAVE": "On Leave",
    "On Leave": "On Leave",
    "OOO": "On Leave"
}

df["status_clean"] = df["status"].map(status_map)

In [ ]:
print("Unmapped statuses:")
print(df.loc[
    df["status"].notna() & df["status_clean"].isna(),
    "status"
].unique())

Unmapped statuses:
[]


In [ ]:
df["status_clean"].value_counts(dropna=False)

,count
status_clean,
Active,2354
Inactive,330
Disabled,239
On Leave,77


for status, I'm intentionally using business meaning, not just formatting normalization That's useful because this master table will eventually be joined to IAM, endpoint, and firewall data. The cybersecurity analysis becomes much more meaningful when we can distinguish an event involving an active employee from one involving someone inactive/disabled.


In [ ]:
print("Hire date examples:")
print(df["hire_date"].dropna().sample(15, random_state=42).to_list())

print("\nTermination date examples:")
print(df["termination_date"].dropna().sample(15, random_state=42).to_list())

Hire date examples:
['26-Oct-2023 17:16:20', '1673981655', '16-Aug-2021 08:48:01', '2022-01-08 21:50:25', '2021-06-17T03:39:43', '2025-09-16 05:30:05', '05/09/2021 21:02', '26/11/2023 06:56', '09-03-2022 11:18:27 PM', '01-01-2023 09:58:56 AM', '06-29-2024 04:03:47 AM', '20/11/2020 23:20', '28-Jun-2023 15:06:00', '2024-06-27 12:27:27', '26/10/2024 15:54']

Termination date examples:
['not available', 'not available', '1727121933', '29-Dec-2022 21:20:26', 'not available', '2021/05/15', '2024-04-23 16:16:28', '04-08-2026 02:23:24 PM', 'not available', 'not available', 'not available', 'not available', 'not available', '23/09/2024', '29/08/2026 11:20']


In [ ]:
print(df["hire_date"].dtype)
print(df["termination_date"].dtype)

object
object


In [ ]:
import pandas as pd
import numpy as np
import re

def clean_mixed_date(value):
    if pd.isna(value):
        return pd.NaT

    value = str(value).strip()

    # Missing indicators
    if value.lower() in ["not available", "na", "n/a", "none", "null", ""]:
        return pd.NaT

    # Unix timestamp (10-digit seconds)
    if re.fullmatch(r"\d{10}", value):
        try:
            return pd.to_datetime(int(value), unit="s")
        except:
            return pd.NaT

    # YYYY-MM-DD / YYYY-MM-DD HH:MM:SS / ISO format
    if re.match(r"^\d{4}-\d{2}-\d{2}", value):
        try:
            return pd.to_datetime(value, dayfirst=False)
        except:
            return pd.NaT

    # YYYY/MM/DD
    if re.match(r"^\d{4}/\d{2}/\d{2}", value):
        try:
            return pd.to_datetime(value, dayfirst=False)
        except:
            return pd.NaT

    # DD/MM/YYYY
    if re.match(r"^\d{2}/\d{2}/\d{4}", value):
        try:
            return pd.to_datetime(value, dayfirst=True)
        except:
            return pd.NaT

    # DD-Mon-YYYY
    if re.match(r"^\d{1,2}-[A-Za-z]{3}-\d{4}", value):
        try:
            return pd.to_datetime(value, dayfirst=True)
        except:
            return pd.NaT

    # MM-DD-YYYY with AM/PM
    if re.match(r"^\d{2}-\d{2}-\d{4}.*(?:AM|PM)$", value, re.IGNORECASE):
        try:
            return pd.to_datetime(value, dayfirst=False)
        except:
            return pd.NaT

    # Anything else — try normal parsing without forcing dayfirst
    try:
        return pd.to_datetime(value, dayfirst=False)
    except:
        return pd.NaT

In [ ]:
df["hire_date_clean"] = df["hire_date"].apply(clean_mixed_date)
df["termination_date_clean"] = df["termination_date"].apply(clean_mixed_date)

In [ ]:
test_values = [
    "2021-11-12 16:30:48",
    "2021-06-17T03:39:43",
    "05/09/2021 21:02",
    "06-29-2024 04:03:47 AM",
    "1673981655",
    "2025/01/01"
]

for x in test_values:
    print(x, " -> ", clean_mixed_date(x))

2021-11-12 16:30:48  ->  2021-11-12 16:30:48
2021-06-17T03:39:43  ->  2021-06-17 03:39:43
05/09/2021 21:02  ->  2021-09-05 21:02:00
06-29-2024 04:03:47 AM  ->  2024-06-29 04:03:47
1673981655  ->  2023-01-17 18:54:15
2025/01/01  ->  2025-01-01 00:00:00


In [ ]:
print(df[[
    "hire_date",
    "hire_date_clean",
    "termination_date",
    "termination_date_clean"
]].sample(15, random_state=42))

                   hire_date     hire_date_clean termination_date  \
1855        16/12/2020 21:52 2020-12-16 21:52:00              NaN   
1211        17/02/2026 19:40 2026-02-17 19:40:00              NaN   
1871  05-13-2026 06:54:37 AM 2026-05-13 06:54:37              NaN   
254      2024-10-13T14:45:26 2024-10-13 14:45:26    not available   
2573              2025/01/01 2025-01-01 00:00:00    not available   
1138  06-02-2026 01:57:44 AM 2026-02-06 01:57:44    not available   
1442     2021-11-12 16:30:48 2021-12-11 16:30:48              NaN   
2174    11-Nov-2021 18:37:43 2021-11-11 18:37:43              NaN   
412   11-10-2025 05:55:15 PM 2025-10-11 17:55:15    not available   
2649              2024/08/25 2024-08-25 00:00:00              NaN   
2316    16-Jun-2020 16:37:07 2020-06-16 16:37:07    not available   
52                1583427464 2020-03-05 16:57:44              NaN   
653      2024-07-26T15:42:06 2024-07-26 15:42:06    not available   
1816        31/08/2020 11:34 2020-

In [ ]:
print(df["hire_date_clean"].dtype)
print(df["termination_date_clean"].dtype)

datetime64[ns]
datetime64[ns]


In [ ]:
print("Hire dates failed to parse:",
      df["hire_date"].notna().sum() - df["hire_date_clean"].notna().sum())

print("Termination dates failed to parse:",
      df["termination_date"].notna().sum() - df["termination_date_clean"].notna().sum())

Hire dates failed to parse: 0
Termination dates failed to parse: 865


In [ ]:
def real_parse_failures(original, cleaned):
    valid_original = (
        original.notna() &
        ~original.astype(str).str.strip().str.lower().isin(
            ["not available", "na", "n/a", "none", "null", ""]
        )
    )

    return (valid_original & cleaned.isna()).sum()

print(
    "Hire date real parse failures:",
    real_parse_failures(df["hire_date"], df["hire_date_clean"])
)

print(
    "Termination date real parse failures:",
    real_parse_failures(df["termination_date"], df["termination_date_clean"])
)

Hire date real parse failures: 0
Termination date real parse failures: 0


In [ ]:
invalid_dates = df[
    df["hire_date_clean"].notna() &
    df["termination_date_clean"].notna() &
    (df["termination_date_clean"] < df["hire_date_clean"])
]

print("Termination before hire:", len(invalid_dates))

Termination before hire: 0


username



In [ ]:
print(df["username"].value_counts(dropna=False).head(30))

username
NaN                       124
  zarna.rao38               1
  nilima.goda53             1
indrajit_palla27            1
chaman.krish7               1
yashawini.warrior76         1
damyanti.ganguly48          1
  ekalinga.sanghvi99        1
warhi.grewal19              1
vrishti.vala48              1
abhimanyu.dhar56            1
max.kant36                  1
yug.lalla64                 1
radha.kibe52                1
lopa.dhawan30               1
yahvi.ramanathan41          1
nimrat.parmer72             1
ethan.sekhon49              1
kalpit_salvi36              1
upkaar.padmanabhan11        1
neel.bhalla45               1
krishna.narang7             1
ekapad.nayar19              1
anamika.deshpande59         1
yochana.patil46             1
idika.rao58                 1
vritti.sengupta40           1
chaman.memon27              1
ranbir.shanker19            1
varenya.nagar50             1
Name: count, dtype: int64


In [ ]:
print("Missing usernames:", df["username"].isna().sum())
print("Unique usernames:", df["username"].nunique())

Missing usernames: 124
Unique usernames: 2876


In [ ]:
df["username_clean"] = (
    df["username"]
    .astype("string")
    .str.strip()
    .str.lower()
)

In [ ]:
username_check = (
    df[df["username_clean"].notna()]
    .groupby("username_clean")["user_id_clean"]
    .nunique()
)

print("Usernames mapped to multiple users:",
      (username_check > 1).sum())

Usernames mapped to multiple users: 0


In [ ]:
print(df[["username", "username_clean"]].dropna().head(20))

                  username       username_clean
0           OMKAAR.CHANA52       omkaar.chana52
1       ethan.narasimhan97   ethan.narasimhan97
2         ethan.choudhry47     ethan.choudhry47
3         vedant_panchal60     vedant_panchal60
4     gayathri.mohanty65     gayathri.mohanty65
5       GARIMA.CHAUDHARY44   garima.chaudhary44
6        vrishti.dewan18        vrishti.dewan18
8             peter.sama77         peter.sama77
9              ayush.dua87          ayush.dua87
10      abhimanyu.trivedi4   abhimanyu.trivedi4
11      christopher.toor37   christopher.toor37
12       urishilla.bhatt88    urishilla.bhatt88
13     GIRINDRA.DHALIWAL46  girindra.dhaliwal46
14      gaurangi.samra66       gaurangi.samra66
15           waida.anand56        waida.anand56
16            LIBNI.LALA75         libni.lala75
17         qabil.lanka74          qabil.lanka74
18             abha.naik17          abha.naik17
19          bimala.raval96       bimala.raval96
20          tristan.ravi53       tristan

hostname

In [ ]:
print(df["hostname"].value_counts(dropna=False).head(30))

hostname
NaN                     171
LPT-11107                 1
vdr-10848                 1
VDR-10966                 1
LPT-12538                 1
LPT-10661                 1
VDR-11002.corp.local      1
WS-12413.corp.local       1
WS-12594                  1
WS-12736                  1
ws-11879                  1
WS-11865                  1
WS-11203                  1
VDR-10024                 1
WS-12895.corp.local       1
WS-12092                  1
VDR-12989                 1
WS-12371                  1
VDR-10875                 1
LPT-12871                 1
WS-10938                  1
WS-12902                  1
WS-12720                  1
LPT-10428                 1
VDR-11328                 1
WS-12064                  1
LPT-10669                 1
vdr-10031                 1
LPT-11341                 1
WS-10722                  1
Name: count, dtype: int64


In [ ]:
print("Missing hostnames:", df["hostname"].isna().sum())
print("Unique hostnames:", df["hostname"].nunique())

Missing hostnames: 171
Unique hostnames: 2829


In [ ]:
print(df["hostname"].dropna().sample(20, random_state=42).to_list())

['WS-10916', 'VDR-12618', 'VDR-10795', 'WS-10334', 'WS-12849', 'LPT_10777', 'LPT-12810', 'WS-12344', 'LPT-11065', 'LPT-10892', 'VDR-10614', 'WS-10831', 'WS-10794', 'WS-12715.corp.local', 'WS-12509', 'LPT-12653', 'WS-12426', 'WS-11567', 'LPT-10842.corp.local', 'vdr-12042']


For joining firewall and endpoint data,  recommended removing .corp.local because the underlying hostname identifier is the WS/VDR/LPT + number.

In [ ]:
#hostname cleaning
import re

def clean_hostname(value):
    if pd.isna(value):
        return pd.NA

    value = str(value).strip().upper()

    # Remove corporate domain suffix
    value = re.sub(r"\.CORP\.LOCAL$", "", value)

    # Standardize underscore to hyphen
    value = value.replace("_", "-")

    return value

df["hostname_clean"] = df["hostname"].apply(clean_hostname)

In [ ]:
print(df[["hostname", "hostname_clean"]].dropna().sample(20, random_state=42))

                  hostname hostname_clean
1111              WS-10916       WS-10916
1402             VDR-12618      VDR-12618
2921             VDR-10795      VDR-10795
1734              WS-10334       WS-10334
211               WS-12849       WS-12849
2006             LPT_10777      LPT-10777
345              LPT-12810      LPT-12810
1405              WS-12344       WS-12344
2534             LPT-11065      LPT-11065
1583             LPT-10892      LPT-10892
1715             VDR-10614      VDR-10614
347               WS-10831       WS-10831
596               WS-10794       WS-10794
1692   WS-12715.corp.local       WS-12715
878               WS-12509       WS-12509
2490             LPT-12653      LPT-12653
2194              WS-12426       WS-12426
1374              WS-11567       WS-11567
863   LPT-10842.corp.local      LPT-10842
1577             vdr-12042      VDR-12042


In [ ]:
print("Missing hostnames:", df["hostname_clean"].isna().sum())
print("Unique cleaned hostnames:", df["hostname_clean"].nunique())

Missing hostnames: 171
Unique cleaned hostnames: 2829


In [ ]:
hostname_check = (
    df[df["hostname_clean"].notna()]
    .groupby("hostname_clean")["user_id_clean"]
    .nunique()
)

print(
    "Hostnames mapped to multiple users:",
    (hostname_check > 1).sum()
)

Hostnames mapped to multiple users: 0


Those 171 records have no hostname in the source data. Cleaning can standardize an existing hostname, but it can't reliably reconstruct a missing one.

So this is actually a data-quality finding we should preserve:

hostname_clean
├── 2,919 records with hostname
└── 171 records missing

And because we already verified:

Hostnames mapped to multiple users: 0

In [ ]:
#######


device_id

In [ ]:
print("Missing device IDs:", df["device_id"].isna().sum())
print("Unique device IDs:", df["device_id"].nunique())

print("\nExamples:")
print(df["device_id"].dropna().sample(20, random_state=42).to_list())

Missing device IDs: 569
Unique device IDs: 2421

Examples:
['dev25568', 'DEV-72126', 'DEV57454', 'DEV-12135', 'dev 32709', 'dev 30027', 'DEV82250', 'dev23006', 'dev 82733', 'dev75773', 'dev21892', 'dev98918', 'dev33784', 'DEV10074', 'dev 27725', 'dev 94926', 'dev62098', 'DEV10766', 'DEV-68390', 'dev 63504']


In [ ]:
def clean_device_id(value):
    if pd.isna(value):
        return pd.NA

    value = str(value).strip().upper()

    # Remove spaces and hyphens
    value = re.sub(r"[\s-]+", "", value)

    return value

df["device_id_clean"] = df["device_id"].apply(clean_device_id)

In [ ]:
print(df[["device_id", "device_id_clean"]].dropna().sample(20, random_state=42))

      device_id device_id_clean
526    dev25568        DEV25568
2393  DEV-72126        DEV72126
756    DEV57454        DEV57454
1373  DEV-12135        DEV12135
2157  dev 32709        DEV32709
2153  dev 30027        DEV30027
1844   DEV82250        DEV82250
2000   dev23006        DEV23006
303   dev 82733        DEV82733
2361   dev75773        DEV75773
423    dev21892        DEV21892
1570   dev98918        DEV98918
256    dev33784        DEV33784
2460   DEV10074        DEV10074
1210  dev 27725        DEV27725
643   dev 94926        DEV94926
465    dev62098        DEV62098
1583   DEV10766        DEV10766
2381  DEV-68390        DEV68390
336   dev 63504        DEV63504


In [ ]:
print("Missing device IDs:", df["device_id_clean"].isna().sum())
print("Unique cleaned device IDs:", df["device_id_clean"].nunique())

Missing device IDs: 569
Unique cleaned device IDs: 2391


In [ ]:
device_check = (
    df[df["device_id_clean"].notna()]
    .groupby("device_id_clean")["user_id_clean"]
    .nunique()
)

print(
    "Device IDs mapped to multiple users:",
    (device_check > 1).sum()
)

Device IDs mapped to multiple users: 39


thats a problem

In [ ]:
device_collisions = (
    df[df["device_id_clean"].notna()]
    .groupby("device_id_clean")["user_id_clean"]
    .nunique()
)

collision_ids = device_collisions[device_collisions > 1].index

df[df["device_id_clean"].isin(collision_ids)][
    ["user_id_clean", "username_clean", "device_id", "device_id_clean",
     "hostname_clean", "department_clean", "status_clean"]
].sort_values("device_id_clean")

,user_id_clean,username_clean,device_id,device_id_clean,hostname_clean,department_clean,status_clean
1930,EMP12459,amaira.dani33,DEV-10071,DEV10071,LPT-12459,Operations,Active
2042,EMP10951,leena.batra60,DEV10071,DEV10071,LPT-10951,Procurement,Active
2757,10016,jacob.ranganathan48,dev 10074,DEV10074,VDR-10016,Finance,Active
2460,EMP12668,manya.chaudry96,DEV10074,DEV10074,WS-12668,Procurement,Active
2628,EMP12521,jackson.sarma53,DEV14911,DEV14911,LPT-12521,Brand,Active
475,EMP12920,lakshit.yohannan97,dev 14911,DEV14911,LPT-12920,Legal,Active
2290,EMP10525,sachi.sinha77,dev 22214,DEV22214,LPT-10525,IT,Active
1630,EMP11825,dipta.bala82,dev22214,DEV22214,LPT-11825,Brand,Active
1922,EMP12873,anusha.sengupta42,DEV23090,DEV23090,WS-12873,Operations,Inactive
223,EMP11511,sudiksha.raman15,DEV23090,DEV23090,LPT-11511,R&D,Active


In [ ]:
device_user_counts = (
    df.groupby("device_id_clean")["user_id_clean"]
    .transform("nunique")
)

df["device_id_conflict"] = (
    df["device_id_clean"].notna() &
    (device_user_counts > 1)
)

In [ ]:
print(df["device_id_conflict"].value_counts())

device_id_conflict
False    2921
True       79
Name: count, dtype: int64


This isn't merely something to hide during cleaning. Shared/conflicting device assignments can themselves be a security signal.

For example, the data contains devices associated with employees from different departments and even different statuses. DEV25765, for instance, appears against three employees, including disabled and inactive users So our rule is:

Normalize → detect conflict → flag → preserve.

Not:

Normalize → delete records.

In [ ]:
print("Conflicting device IDs:",
      df.loc[df["device_id_conflict"], "device_id_clean"].nunique())

print("Records affected:",
      df["device_id_conflict"].sum())

Conflicting device IDs: 39
Records affected: 79


In [ ]:
df["device_user_count"] = (
    df.groupby("device_id_clean")["user_id_clean"]
    .transform("nunique")
)

manager_username

In [ ]:
print("Missing manager usernames:", df["manager_username"].isna().sum())
print("Unique manager usernames:", df["manager_username"].nunique())

print("\nExamples:")
print(df["manager_username"].dropna().sample(25, random_state=42).to_list())

Missing manager usernames: 637
Unique manager usernames: 1165

Examples:
['manager188', 'manager647', 'manager140', 'manager402', 'manager178', 'manager416', 'manager903', 'manager866', 'manager453', 'manager546', 'manager897', 'MANAGER874', 'manager421', 'manager111', 'manager132', 'manager795', 'MANAGER546', 'manager215', 'manager815', 'manager270', 'manager253', 'manager730', 'manager736', 'manager850', 'manager340']


In [ ]:
df["manager_username_clean"] = (
    df["manager_username"]
    .astype("string")
    .str.strip()
    .str.lower()
)

In [ ]:
print(df[[
    "manager_username",
    "manager_username_clean"
]].dropna().sample(20, random_state=42))

     manager_username manager_username_clean
1500       manager188             manager188
2136       manager647             manager647
2109       manager140             manager140
2589       manager402             manager402
2007       manager178             manager178
280        manager416             manager416
1843       manager903             manager903
327        manager866             manager866
2510       manager453             manager453
840        manager546             manager546
1551       manager897             manager897
1848       MANAGER874             manager874
2116       manager421             manager421
908        manager111             manager111
2903       manager132             manager132
2650       manager795             manager795
2023       MANAGER546             manager546
2969       manager215             manager215
2945       manager815             manager815
776        manager270             manager270


In [ ]:
print("Missing manager usernames:",
      df["manager_username_clean"].isna().sum())

print("Unique cleaned manager usernames:",
      df["manager_username_clean"].nunique())

Missing manager usernames: 637
Unique cleaned manager usernames: 821


In [ ]:
employee_usernames = set(
    df["username_clean"].dropna()
)

manager_usernames = set(
    df["manager_username_clean"].dropna()
)

managers_not_found = manager_usernames - employee_usernames

print("Managers not found in employee usernames:",
      len(managers_not_found))

print("Examples:")
print(list(managers_not_found)[:20])

Managers not found in employee usernames: 821
Examples:
['manager154', 'manager322', 'manager308', 'manager608', 'manager137', 'manager251', 'manager391', 'manager362', 'manager230', 'manager887', 'manager807', 'manager124', 'manager529', 'manager423', 'manager690', 'manager393', 'manager289', 'manager928', 'manager687', 'manager935']


In [ ]:
df["manager_username_clean"] = (
    df["manager_username"]
    .astype("string")
    .str.strip()
    .str.lower()
)

In [ ]:
identity_master_clean = df.copy()

print("Final shape:", identity_master_clean.shape)
print("Duplicate rows:", identity_master_clean.duplicated().sum())
print("Duplicate user IDs:", identity_master_clean["user_id_clean"].duplicated().sum())

Final shape: (3000, 24)
Duplicate rows: 0
Duplicate user IDs: 0


In [ ]:
print("=== IDENTITY MASTER QUALITY AUDIT ===")

print("\nShape:")
print(identity_master_clean.shape)

print("\nDuplicate rows:")
print(identity_master_clean.duplicated().sum())

print("\nDuplicate user IDs:")
print(identity_master_clean["user_id_clean"].duplicated().sum())

print("\nMissing values:")
print(
    identity_master_clean[
        [
            "user_id_clean",
            "username_clean",
            "department_clean",
            "location_clean",
            "hostname_clean",
            "device_id_clean",
            "manager_username_clean",
            "hire_date_clean",
            "termination_date_clean"
        ]
    ].isna().sum()
)

print("\nDevice conflicts:")
print(identity_master_clean["device_id_conflict"].sum())

print("\nDate validation:")
print(
    (
        identity_master_clean["termination_date_clean"].notna()
        & identity_master_clean["hire_date_clean"].notna()
        & (
            identity_master_clean["termination_date_clean"]
            < identity_master_clean["hire_date_clean"]
        )
    ).sum(),
    "termination dates before hire dates"
)

=== IDENTITY MASTER QUALITY AUDIT ===

Shape:
(3000, 24)

Duplicate rows:
0

Duplicate user IDs:
0

Missing values:
user_id_clean                0
username_clean             124
department_clean             0
location_clean             564
hostname_clean             171
device_id_clean            569
manager_username_clean     637
hire_date_clean            329
termination_date_clean    2631
dtype: int64

Device conflicts:
79

Date validation:
0 termination dates before hire dates
